src/input.js

```diff
@@ -58,35 +64,186 @@ function maybeSetSelection(state, dispatch, selection) {
   return true;
 }
 
+function withFlushedState(view, state, f) {
+  const viewState = view.state,
+    active = view.root.activeElement;
+  if (viewState != state) view.updateState(state);
+  if (active != view.dom) view.focus();
+  try {
+    return f();
+  } finally {
+    if (viewState != state) view.updateState(viewState);
+    if (active != view.dom && active) active.focus();
+  }
+}
+
+let reusedRange = null;
+
+// Note that this will always return the same range, because DOM range
+// objects are every expensive, and keep slowing down subsequent DOM
+// updates, for some reason.
+export const textRange = function (node, from, to) {
+  const range = reusedRange || (reusedRange = document.createRange());
+  range.setEnd(node, to == null ? node.nodeValue.length : to);
+  range.setStart(node, from || 0);
+  return range;
+};
+
+// : (EditorView, number, number)
+// Whether vertical position motion in a given direction
+// from a position would leave a text block.
+function endOfTextblockVertical(view, state, dir, selection) {
+  const sel = selection;
+  const $pos = dir === 'up' ? sel.$from : sel.$to;
+  return withFlushedState(view, state, () => {
+    let {node: dom} = view.docView.domFromPos($pos.pos, dir == 'up' ? -1 : 1);
+    for (;;) {
```

Comment:
how does this work?

Reply:
same here

Reply:
just wondering how this for loop `for (;;)` works?

Reply:
is this an infinite loop? @ehsanmasdar

Reply:
yes 👍🏽